test/e2e/azure_securitygroups.go

```diff
@@ -0,0 +1,187 @@
+			Source:           pointer.String("*"),
+			Destination:      pointer.String("*"),
+		}
+		testSecurityRule2 = infrav1.SecurityRule{
+			Name:             "test-security-rule-2",
+			Description:      "test-security-rule-2",
+			Protocol:         "Tcp",
+			Direction:        "Inbound",
+			Priority:         110,
+			SourcePorts:      pointer.String("*"),
+			DestinationPorts: pointer.String("80"),
+			Source:           pointer.String("*"),
+			Destination:      pointer.String("*"),
+		}
+		input AzureSecurityGroupsSpecInput
+	)
+
+	Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	Expect(input.Namespace).ToNot(BeNil(), "Invalid argument. input.Namespace can't be nil when calling %s spec", specName)
+	Expect(input.ClusterName).ToNot(BeEmpty(), "Invalid argument. input.ClusterName can't be empty when calling %s spec", specName)
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, input.Namespace.Name, input.ClusterName)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	// get subscription id
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := azureutil.GetAuthorizer(settings)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("creating a subnets client")
+	subnetsClient := network.NewSubnetsClient(subscriptionID)
+	subnetsClient.Authorizer = auth
+
+	azureCluster := &infrav1.AzureCluster{}
+	err = mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.InfrastructureRef.Namespace,
+		Name:      input.Cluster.Spec.InfrastructureRef.Name,
+	}, azureCluster)
+	Expect(err).NotTo(HaveOccurred())
+
+	var expectedSubnets infrav1.Subnets
+	checkSubnets := func(g Gomega) {
+		actualSubnetsListResult, err := subnetsClient.List(ctx, azureCluster.Spec.ResourceGroup, azureCluster.Spec.NetworkSpec.Vnet.Name)
+		Expect(err).NotTo(HaveOccurred())
+		actualSubnets := actualSubnetsListResult.Values()
+		Logf("actualSubnets: %v", actualSubnets)
+
+		var trackedSubnets []*network.Subnet
+		for _, subnet := range actualSubnets {
+			subnet := subnet
+			if strings.HasPrefix(*subnet.Name, "test") {
+				trackedSubnets = append(trackedSubnets, &subnet)
+			}
+		}
+
+		if len(trackedSubnets) == 0 {
+			trackedSubnets = nil
+		}
+		if expectedSubnets == nil {
+			g.Expect(trackedSubnets).To(BeNil())
+		} else {
+			g.Expect(len(trackedSubnets)).To(Equal(len(expectedSubnets)))
+			for i := range expectedSubnets {
+				g.Expect(trackedSubnets[i].SubnetPropertiesFormat.NetworkSecurityGroup.SecurityGroupPropertiesFormat.SecurityRules).To(Equal(expectedSubnets[i].SecurityGroup.SecurityRules))
```

Comment:
It looks like this test is failing in CI and I'm guessing it's at least in part because these actual and expected values are different types.